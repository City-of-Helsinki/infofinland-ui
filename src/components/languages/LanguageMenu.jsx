import Link from 'next/link'
import { IconCheck, IconGlobe } from '../Icons'
import Drawer from '../layout/Drawer'
import { useRouter } from 'next/router'
import cls from 'classnames'
import useTranslation from 'next-translate/useTranslation'

import { languages } from '../../../i18n'

export const LanguageMenu = ({ closeMenu }) => {
  const { t } = useTranslation('common')

  const { pathname, query, locale } = useRouter()
  return (
    <div role="listbox">
      <label className="block px-14 mb-8 text-body-large font-bold">
        {t('languageMenu.label')}
      </label>
      {languages.map((lang) => {
        const { text, code } = lang
        const isSelected = locale === code
        return (
          <div
            onClick={closeMenu}
            key={`lang-${code}`}
            className={cls({
              'font-bold': isSelected,
            })}
          >
            <Link
              passHref
              href={{ pathname, query }}
              locale={code}
              scroll={false}
            >
              <a
                role="option"
                aria-selected={code === locale}
                className="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
                title={text}
                lang={code}
                hrefLang={code}
              >
                <span className="inline-block w-16 text-body-small font-bold text-gray-medium uppercase text-bodytext-color-op5 pe-8 float-start">
                  {code}
                </span>
                <span className="inline-block text-body-small text-bodytext-color float-start">
                  {text}
                </span>
                {isSelected && <IconCheck className="ms-4" />}
              </a>
            </Link>
          </div>
        )
      })}
    </div>
  )
}

export const LangMenuDrawer = ({ closeMenu, isOpen }) => (
  <Drawer close={closeMenu} isOpen={isOpen}>
    <LanguageMenu closeMenu={closeMenu} />
  </Drawer>
)

export const LanguageMenuButton = ({ onClick }) => {
  const { t } = useTranslation('common')
  const { locale } = useRouter()
  // const { text } = languages.find(({ code }) => code === locale)
  return (
    <button
      title={t('languageMenu.label')}
      className=" block lg:hidden h-8 transform -translate-y-0.5 md:me-2"
      onClick={onClick}
    >
      <span className=" inline-block text-action uppercase">{locale}</span>
      {/* <span className=" hidden md:inline-block text-action uppercase">
        {text}
      </span> */}
      <IconGlobe className="mx-2 xl:mx-0 w-4 md:w-5 h-4 md:h-5" />
    </button>
  )
}
