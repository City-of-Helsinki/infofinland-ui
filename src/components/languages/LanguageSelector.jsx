import { useRouter } from 'next/router'
import Link from 'next/link'
import { IconGlobe } from '@/components/Icons'
import cls from 'classnames'
import { useTranslation } from 'next-i18next'
import { i18n } from '@/next-i18next.config'
import { getLocalesForPath } from '@/lib/client-api'
import useSWR from 'swr'

const LanguageSelector = ({ openMenu }) => {
  const { locale, asPath: path } = useRouter()
  const { t } = useTranslation('common')
  const cacheKey = path ? path : null
  const fetcher = () => getLocalesForPath({ path })
  const { data: locales, error } = useSWR(cacheKey, fetcher)

  return (
    <>
      <div
        dir="ltr"
        className=" hidden lg:block text-action leading-10 divide-x xl:divide-x-0 divide-black me-4 md:ms-1"
      >
        <button
          className=" xl:hidden lg:px-0 pe-2 ps-2"
          onClick={openMenu}
          aria-haspopup="dialog"
        >
          <span className="hidden md:inline-block xl:hidden">
            {t('languageSelector.button')}
          </span>
          <IconGlobe className="xl:hidden mx-2 xl:mx-0 w-5 h-5" />
        </button>
        {i18n.languages.map(({ text, code }) => {
          const isLocalized =
            locales?.find((page) => page.locale === code) !== undefined

          return (
            <Link
              href={path}
              locale={code}
              passHref
              scroll={false}
              // prefetch={false}
              key={`lang-${code}`}
            >
              <a
                className={cls(
                  'xl:inline-block px-3 2xl:px-4 font-bold text-center uppercase ',
                  {
                    'text-black': error || !locales || isLocalized,
                    'text-gray': locales?.length > 0 && !isLocalized && !error,
                  }
                )}
                title={text}
                hrefLang={code}
                lang={code}
              >
                <span
                  className={cls('xl:hidden', {
                    ' border-b-2 border-black': locale === code,
                  })}
                >
                  {code}
                </span>
                <span
                  className={cls(
                    'hidden xl:inline-block hover:border-black border-b-2 transition-all ease duration-150',
                    {
                      'border-black': locale === code,
                      'border-white': locale !== code,
                    }
                  )}
                >
                  {text}
                </span>
              </a>
            </Link>
          )
        })}
      </div>
    </>
  )
}

export default LanguageSelector
