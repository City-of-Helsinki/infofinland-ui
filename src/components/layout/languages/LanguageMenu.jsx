import Link from 'next/link'
import { IconCheck, IconGlobe } from '@/components/Icons'
import Drawer from '@/components/layout/Drawer'
import { useRouter } from 'next/router'
import cls from 'classnames'

import { useTranslation } from 'next-i18next'

import { i18n } from '@/next-i18next.config'

//For accessibility and to avoid over-engineering and excessive imports,
// Label translations are copied here for multi lang use.
const LABELS = {
  fi: 'Valitse kieli',
  sv: 'Välj språk',
  en: 'Select language',
  ru: 'Выбрать язык',
  et: 'Vali keel',
  fr: 'Sélectionner la langue',
  so: 'Dooro luqadda',
  es: 'Elegir idioma',
  tr: 'Dil seçin',
  zh: '选择语言',
  fa: 'زبان را انتخاب کن',
  ar: 'اختر اللغة',
}

export const LanguageMenu = ({ closeMenu }) => {
  const { t } = useTranslation('common')
  const { asPath, locale } = useRouter()
  return (
    <div>
      <label className="block px-14 mb-8 text-body-large font-bold">
        {t('languageMenu.label')}
      </label>
      {i18n.supportedLanguages.map((lang) => {
        const { text, code } = lang
        const isSelected = locale === code
        return (
          <div
            onClick={closeMenu}
            key={`lang-${code}`}
            className={cls({
              'font-bold': isSelected,
            })}
          >
            <Link
              passHref
              href={asPath}
              locale={code}
              scroll={false}
              prefetch={false}
            >
              <a
                className={cls(
                  'flex items-center py-2 px-14 hover:bg-gray-white clear-start',
                  {
                    // 'system-zh-font':code === 'zh'
                  }
                )}
                title={text}
                lang={code}
                hrefLang={code}
              >
                <span className="inline-block text-body-small text-bodytext-color float-start">
                  {text}
                </span>
                {isSelected && <IconCheck className="ms-4" />}
              </a>
            </Link>
          </div>
        )
      })}
    </div>
  )
}

export const LangMenuDrawer = ({ closeMenu, isOpen }) => (
  <Drawer close={closeMenu} isOpen={isOpen}>
    <LanguageMenu closeMenu={closeMenu} />
  </Drawer>
)

export const LanguageMenuButton = ({ onClick }) => {
  const { t } = useTranslation('common')
  let { locale } = useRouter()
  if (i18n.disabledLocales.includes(locale)) {
    locale = i18n.defaultLocale
  }

  const { text } = i18n.supportedLanguages.find(({ code }) => code === locale)
  const assisted = i18n.supportedLanguages.filter(({ code }) => code !== locale)
  return (
    <button
      aria-haspopup="dialog"
      title={t('languageMenu.label')}
      className=" flex lg:hidden items-center h-10 md:me-2"
      onClick={onClick}
    >
      <span className="text-action uppercase">{text}</span>
      {assisted.map(({ code }) => (
        <span className="hidden" key={`assist-label-${code}`} lang={code}>
          {LABELS[code]}
        </span>
      ))}
      <IconGlobe className="mx-2 xl:mx-0 w-4 md:w-5 md:h-5" />
    </button>
  )
}
