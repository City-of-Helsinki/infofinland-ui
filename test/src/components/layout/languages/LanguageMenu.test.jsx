/**
 * @jest-environment jsdom
 */
/* eslint-disable import/no-extraneous-dependencies */
import { render } from '@testing-library/react'
import { LanguageMenu, LanguageMenuButton } from '@/components/layout/languages/LanguageMenu'

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    locale: 'fi',
    asPath: '/foo/bar',
    events: {
      on: jest.fn(),
      off: jest.fn()
    },
  })
}))

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return {
      t: (i18nKey) => i18nKey,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));

describe('LanguageMenu', () => {
  it('should render the LanguageMenu component', () => {
    const { asFragment } = render(<LanguageMenu />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render the LanguageMenuButton component', () => {
    const { asFragment } = render(<LanguageMenuButton />)
    expect(asFragment()).toMatchSnapshot()
  })
})
