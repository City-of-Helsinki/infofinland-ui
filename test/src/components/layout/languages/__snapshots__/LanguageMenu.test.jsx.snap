// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LanguageMenu should render the LanguageMenu component 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block px-14 mb-8 text-body-large font-bold"
    >
      languageMenu.label
    </label>
    <div
      class="font-bold"
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="fi"
        lang="fi"
        title="Suomi"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Suomi
        </span>
        <svg
          class="inline-block ms-4"
          fill="none"
          height="11"
          viewBox="0 0 14 11"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13 1L4.99967 9L1 4.9995"
            stroke="black"
            stroke-width="2"
          />
        </svg>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="sv"
        lang="sv"
        title="Svenska"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Svenska
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="en"
        lang="en"
        title="English"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          English
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="ru"
        lang="ru"
        title="Pусский"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Pусский
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="et"
        lang="et"
        title="Eesti keel"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Eesti keel
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="uk"
        lang="uk"
        title="Українська"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Українська
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="fr"
        lang="fr"
        title="Français"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Français
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="es"
        lang="es"
        title="Español"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Español
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="tr"
        lang="tr"
        title="Türkçe"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          Türkçe
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="zh"
        lang="zh"
        title="中文"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          中文
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="fa"
        lang="fa"
        title=" فارسی"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
           فارسی
        </span>
      </a>
    </div>
    <div
      class=""
    >
      <a
        class="flex items-center py-2 px-14 hover:bg-gray-white clear-start"
        href="/foo/bar"
        hreflang="ar"
        lang="ar"
        title="العربية"
      >
        <span
          class="inline-block text-body-small text-bodytext-color float-start"
        >
          العربية
        </span>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LanguageMenu should render the LanguageMenuButton component 1`] = `
<DocumentFragment>
  <button
    aria-haspopup="dialog"
    class=" flex lg:hidden items-center h-10 md:me-2"
    title="languageMenu.label"
  >
    <span
      class="text-action uppercase"
    >
      Suomi
    </span>
    <span
      class="hidden"
      lang="sv"
    >
      Välj språk
    </span>
    <span
      class="hidden"
      lang="en"
    >
      Select language
    </span>
    <span
      class="hidden"
      lang="ru"
    >
      Выбрать язык
    </span>
    <span
      class="hidden"
      lang="et"
    >
      Vali keel
    </span>
    <span
      class="hidden"
      lang="uk"
    />
    <span
      class="hidden"
      lang="fr"
    >
      Sélectionner la langue
    </span>
    <span
      class="hidden"
      lang="es"
    >
      Elegir idioma
    </span>
    <span
      class="hidden"
      lang="tr"
    >
      Dil seçin
    </span>
    <span
      class="hidden"
      lang="zh"
    >
      选择语言
    </span>
    <span
      class="hidden"
      lang="fa"
    >
      زبان را انتخاب کن
    </span>
    <span
      class="hidden"
      lang="ar"
    >
      اختر اللغة
    </span>
    <svg
      class="inline-block mx-2 xl:mx-0 w-4 md:w-5 md:h-5"
      viewBox="0 0 200 200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M100,175.69A75.69,75.69,0,1,1,175.69,100,75.78,75.78,0,0,1,100,175.69Zm0-141.38A65.69,65.69,0,1,0,165.69,100,65.77,65.77,0,0,0,100,34.31Z"
        fill="black"
      />
      <path
        d="M100,175.69c-19.49,0-30-39-30-75.69s10.51-75.69,30-75.69,30,39,30,75.69S119.49,175.69,100,175.69Zm0-141.38c-8.15,0-20,25.59-20,65.69,0,18.33,2.5,35.49,7,48.32,3.73,10.55,8.82,17.37,13,17.37,8.15,0,20-25.59,20-65.69,0-18.33-2.5-35.49-7-48.32C109.23,41.13,104.14,34.31,100,34.31Z"
        fill="black"
      />
      <path
        d="M170.69,105H29.31a5,5,0,0,1,0-10H170.69a5,5,0,0,1,0,10Z"
        fill="black"
      />
      <path
        d="M163.05,73H37a5,5,0,0,1,0-10h126.1a5,5,0,0,1,0,10Z"
        fill="black"
      />
      <path
        d="M163.05,137H37a5,5,0,0,1,0-10h126.1a5,5,0,0,1,0,10Z"
        fill="black"
      />
    </svg>
  </button>
</DocumentFragment>
`;
